components/support/jwcrypto/src/lib.rs

```diff
@@ -176,17 +196,85 @@ impl ToString for CompactJwe {
 /// Encrypt and serialize data in the JWE compact form.
 pub fn encrypt_to_jwe(data: &[u8], encryption_params: EncryptionParameters) -> Result<String> {
     let jwe = match encryption_params {
-        EncryptionParameters::ECDH_ES { .. } => ec::encrypt_to_jwe(data, encryption_params)?,
-    };
+        EncryptionParameters::ECDH_ES { enc, peer_jwk } => ec::encrypt_to_jwe(data, enc, peer_jwk),
+        EncryptionParameters::Direct { jwk } => direct::encrypt_to_jwe(data, jwk),
+    }?;
     Ok(jwe.to_string())
 }
 
 /// Deserialize and decrypt data in the JWE compact form.
 pub fn decrypt_jwe(jwe: &str, decryption_params: DecryptionParameters) -> Result<String> {
     let jwe = jwe.parse()?;
-    Ok(match decryption_params {
-        DecryptionParameters::ECDH_ES { .. } => ec::decrypt_jwe(&jwe, decryption_params)?,
-    })
+    match decryption_params {
+        DecryptionParameters::ECDH_ES { local_key_pair } => ec::decrypt_jwe(&jwe, local_key_pair),
+        DecryptionParameters::Direct { jwk } => direct::decrypt_jwe(&jwe, jwk),
+    }
+}
+
+// raw aes helpers - should move them?
+pub(crate) fn aes_gcm_encrypt(
+    data: &[u8],
+    protected_header: JweHeader,
+    secret: &[u8],
+) -> Result<CompactJwe> {
+    let encryption_algorithm = match protected_header.enc {
+        EncryptionAlgorithm::A256GCM => &aead::AES_256_GCM,
+    };
+    let sealing_key = aead::SealingKey::new(encryption_algorithm, secret)?;
+    let additional_data = serde_json::to_string(&protected_header)?;
+    let additional_data =
+        base64::encode_config(additional_data.as_bytes(), base64::URL_SAFE_NO_PAD);
+    let additional_data = additional_data.as_bytes();
+    let aad = aead::Aad::from(additional_data);
+    let mut iv: Vec<u8> = vec![0; 12];
+    rand::fill(&mut iv)?;
+    let nonce = aead::Nonce::try_assume_unique_for_key(encryption_algorithm, &iv)?;
+    let mut encrypted = aead::seal(&sealing_key, nonce, aad, data)?;
+
+    let tag_idx = encrypted.len() - encryption_algorithm.tag_len();
+    let auth_tag = encrypted.split_off(tag_idx);
+    let ciphertext = encrypted;
+
+    CompactJwe::new(
+        Some(protected_header),
+        None,
+        Some(iv),
+        ciphertext,
+        Some(auth_tag),
+    )
+}
+
+pub(crate) fn aes_gcm_decrypt(
+    jwe: &CompactJwe,
+    protected_header: JweHeader,
+    secret: &[u8],
+) -> Result<String> {
+    // Part 2: decrypt the payload
+    if jwe.encrypted_key()?.is_some() {
+        return Err(JwCryptoError::IllegalState(
+            "The Encrypted Key must be empty.",
+        ));
+    }
+    let encryption_algorithm = match protected_header.enc {
+        EncryptionAlgorithm::A256GCM => &aead::AES_256_GCM,
+    };
```

Comment:
Ah, I see. So IMHO this line should be reading the protected header from the provided `jwe` rather than accepting it as a separate parameter. Accepting it separately only creates opportunities for calling code to get confused and provide the wrong header and open itself up to cryptographic mischief. I think you should remove the `protected_header` parameter from this function entirely.